nilearn/_utils/niimg_conversions.py

```diff
@@ -372,7 +376,21 @@ def check_niimgs(niimgs, accept_3d=False):
 
     # Now, we load data as we know its format
     if dim == 4:
-        niimg = check_niimg(niimgs)
+        if return_iterator:
+            result = (_index_niimgs(niimgs, i)
+                      for i in range(_get_shape(niimgs)[3]))
```

Comment:
> Isn't that suboptimal if niimgs is a 4d nifti?

Reminde me what did you mean by suboptimal? I seem to remember that you were concerned that `check_niimg` will be called on individual 3d images but that's not the case because `_index_nimgs` doesn't include checks and only does the data slicing + 3d image creation.